frontend/src/app/utils/timeline/timeline-creator.ts

```diff
@@ -0,0 +1,296 @@
+        let timeline: Timeline = {
+            columnNumber: this.columnNumber,
+            pageNumber,
+            tasksByRows: [],
+            exceedingTaskBlocks: [],
+            boundaries: this.boundaries,
+            sections: this.getTimelineSections(this.boundaries, scale),
+            subsections: this.getTimelineSubsections(this.boundaries, scale)
+        };
+        let tasksToAdd = this.getVisibleTasks(tasks).map(task => this.addColumns(task));
+        timeline = this.addTasks(tasksToAdd, timeline);
+        return timeline;
+    }
+
+    public getDateRatio(date: Date): number {
+        return (date.getTime() - this.boundaries.start.getTime()) / this.getTimelineLength();
+    }
+
+    private addColumns(task: Task): Task {
+        return { ...task, startColumn: this.getColumn(task.start), endColumn: this.getColumn(task.end) };
+    }
+
+    private getVisibleTasks(tasks: Task[]): Task[] {
+        return tasks.filter(t => this.isVisible(t));
+    }
+
+    private isVisible(task: Task): boolean {
+        return this.getColumn(task.end) - this.getColumn(task.start) >= this.minimumColumnNumber;
+    }
+
+    private getColumn(date: Date): number {
+        if (date < this.boundaries.start) {
+            return 1;
+        }
+        if (date > this.boundaries.end) {
+            return this.columnNumber + 1;
+        }
+        return Math.round(this.getDateRatio(date) * this.columnNumber) + 1;
+    }
+
+    private getBlockStart(block: Task[]): Date {
+        return block.sort((first, second) => first.start.getTime() - second.start.getTime())[0].start;
+    }
+
+    private getBlockEnd(block: Task[]): Date {
+        return block.sort((first, second) => second.end.getTime() - first.end.getTime())[0].end;
+    }
+
+    private getTimelineLength(): number {
+        return this.boundaries.end.getTime() - this.boundaries.start.getTime();
+    }
+
+    private addTasks(tasks: Task[], timeline: Timeline): Timeline {
+        let tasksByRows: Task[][] = Array.from(Array(this.taskRowsNumber), () => []);
+        let exceeding: Task[][] = [];
+        for (let task of tasks) {
+            let inserted = false;
+            for (let i = 0; i < tasksByRows.length && !inserted; i++) {
+                if (!tasksByRows[i].find(t => intercept(t, task))) {
+                    tasksByRows[i].push(task);
+                    inserted = true;
+                }
+            }
+            if (!inserted) {
+                this.addToExceeding(exceeding, task);
+            }
+        }
+        let exceedingTaskBlocks: TasksBlock[] = [];
+        if (exceeding.length > 0) {
+            exceedingTaskBlocks = this.getExceedingBlocks(exceeding);
+        }
+        return { ...timeline, tasksByRows, exceedingTaskBlocks };
+    }
+
+    private getExceedingBlocks(exceeding: Task[][]): TasksBlock[] {
+        let exceedingTaskBlocks: Task[][] = [];
+        exceeding = exceeding.sort(
```

Comment:
Avoid reassigning of function parameters. it may lead to unexpected mutations if you are not careful. Especially in Angular NgRx world where you objects must stay immutable in store it will lead to bad side effects.

You may notice it better when you change your code style to work more with Observables (it will be in future PRs).

https://medium.com/the-andela-way/its-bad-practice-to-reassign-function-params-c37f289a5dc4

Reply:
Fixed